src/hooks/useQueryState.js

```diff
@@ -0,0 +1,163 @@
+   * Comparison of the two effects: https://kentcdodds.com/blog/useeffect-vs-uselayouteffect
+   */
+  useLayoutEffect(() => {
+    const loadURLParams = async () => {
+      const urlParams = new URLSearchParams(history.location.search);
+      let urlValue = urlParams.get(key);
+
+      switch (type) {
+        // using boolean key -> load true if present, false if absent from url
+        case types.Boolean:
+          setState(urlParams.has(key));
+          return;
+
+        // different logic if we are working with an array than with single value
+        // because setState will not do deep array comparison
+        case types.Array:
+          // if the same value (or unset and state array is empty) -> do nothing
+          // this check is necessary because [] -> [] are different memory objects
+          if (urlValue === state.toString() || (!urlValue && state.length === 0)) return;
+          else if (!urlValue) setState(defaultVal);
+          else setState(urlValue.split(','));
+          break;
+
+        case types.SingleValue:
+          setState(urlValue ?? defaultVal);
+          break;
+
+        default:
+          // should never be reached, type must always be one of the preceeding
+          throw new Error(`missing case for type ${type}`);
+      }
+    };
+    loadURLParams();
+    /*********************** ESLint Disable Justification: *******************************/
+    // ESLint wants to include `defaultVal, key, state` as dependencies but to prevent
+    // extra loops of ~ url setting state setting the url ~ we do not include.
+    // It makes good sense that the only time we should pull the URL is the URL changes
+    /*************************************************************************************/
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [history.location.search]);
+
+  /**
+   * Callback function to allow user component to setQueryState just like one would use setState
+   */
+  const setQueryState = useCallback((val) => {
+    // this check prevents unnecessary array setting
+    // ([] -> [] are technically different memory objects)
+    // this is safe to deep compare because the array value should not have very many values
+    // (ex. 'filters' array has a few different possible filter values)
+    if (Array.isArray(defaultVal) && !JSON.stringify(state) === JSON.stringify(val)) return;
+    setState(val);
+    /*********************** ESLint Disable Justification: *******************************/
+    // We do not want the defaultVal or state dependencies here because they were simply added to
+    // prevent unecessarily setting the value array. We do not want any dependencies that might
+    // change the state after the initial setQueryState is called from the user function.
+    /*************************************************************************************/
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
```

Comment:
I think you are misunderstanding how `useCallback` works. Unlike `useEffect`, we do not call the callback of `useCallback` when its dependencies change. Rather, what happens is the callback returned by `useCallback` has its definition updated. The way you have this defined currently, `useCallback` will create the `setQueryState` with `state` = `initial ?? defaultVal` (the value returned by `useState` above). That will never change. So even as `state` gets updated over time, `setQueryState` will still be comparing `val` to that first state of `initial ?? defaultVal`. I'm fairly certain the intended behavior is that `setQueryState` is redefined whenever `state` changes so that it is comparing against the most recent value of `state`.

Also, since `defaultVal` is stable, there should be no issues including it in the dependency array.